packages/core/content-manager/admin/src/preview/utils/previewScript.ts

```diff
@@ -23,9 +33,212 @@ const previewScript = (shouldRun = true) => {
+        );
+
+        highlights.push(highlight);
+        overlay.appendChild(highlight);
+
+        drawOverlay(element, highlight);
+      }
+    });
+
+    return {
+      elements,
+      updateAllHighlights,
+      eventListeners,
+    };
+  };
+
+  type HighlightManager = ReturnType<typeof createHighlightManager>;
+
+  const setupObservers = (highlightManager: HighlightManager) => {
+    const resizeObserver = new ResizeObserver(() => {
+      highlightManager.updateAllHighlights();
+    });
+
+    highlightManager.elements.forEach((element: Element) => {
+      resizeObserver.observe(element);
+    });
+
+    resizeObserver.observe(document.documentElement);
+
+    const updateOnScroll = () => {
+      highlightManager.updateAllHighlights();
+    };
+
+    // Find all scrollable ancestors for all tracked elements
+    const scrollableElements = new Set<Element | Window>();
+    scrollableElements.add(window); // Add window as a special case
+
+    highlightManager.elements.forEach((element) => {
+      let parent = element.parentElement;
+      while (parent) {
+        const computedStyle = window.getComputedStyle(parent);
+        const overflow = computedStyle.overflow + computedStyle.overflowX + computedStyle.overflowY;
+
+        if (overflow.includes('scroll') || overflow.includes('auto')) {
+          scrollableElements.add(parent);
+        }
+
+        parent = parent.parentElement;
+      }
+    });
+
+    // Add scroll listeners to all scrollable elements
+    scrollableElements.forEach((element) => {
+      if (element === window) {
+        window.addEventListener('scroll', updateOnScroll);
+        window.addEventListener('resize', updateOnScroll);
+      } else {
+        (element as Element).addEventListener('scroll', updateOnScroll);
+      }
+    });
+
+    return {
+      resizeObserver,
+      updateOnScroll,
+      scrollableElements,
+    };
+  };
+
+  const setupEventHandlers = (highlightManager: HighlightManager) => {
+    // TODO: The listeners for postMessage events will go here
+    return highlightManager.eventListeners;
+  };
+
+  const createCleanupSystem = (
+    overlay: HTMLElement,
+    observers: ReturnType<typeof setupObservers>,
+    eventHandlers: EventListenersList
+  ) => {
+    window.__strapi_previewCleanup = () => {
+      observers.resizeObserver.disconnect();
+
+      // Remove all scroll listeners
+      observers.scrollableElements.forEach((element) => {
+        if (element === window) {
+          window.removeEventListener('scroll', observers.updateOnScroll);
+          window.removeEventListener('resize', observers.updateOnScroll);
+        } else {
+          (element as Element).removeEventListener('scroll', observers.updateOnScroll);
+        }
+      });
+
+      // Remove highlight event listeners
+      eventHandlers.forEach(({ element, type, handler }) => {
+        element.removeEventListener(type, handler);
+      });
+
+      overlay.remove();
```

Comment:
I know we said the other days that all overlays could be moved up at the body root instead of in a .overlay div. But personally I like the div because I can remove everything at once